modal/io_streams.py

```diff
@@ -245,27 +246,33 @@ def write_eof(self):
 
     async def drain(self):
         """
-        Flushes the write buffer and EOF to the running process.
+        Flushes the write buffer to the running process. Flushes the EOF if the writer is closed.
         """
         data = bytes(self._buffer)
         self._buffer.clear()
         index = self.get_next_index()
 
-        if self._object_type == "sandbox":
-            await retry_transient_errors(
-                self._client.stub.SandboxStdinWrite,
-                api_pb2.SandboxStdinWriteRequest(
-                    sandbox_id=self._object_id, index=index, eof=self._is_closed, input=data
-                ),
-            )
-        else:
-            await retry_transient_errors(
-                self._client.stub.ContainerExecPutInput,
-                api_pb2.ContainerExecPutInputRequest(
-                    exec_id=self._object_id,
-                    input=api_pb2.RuntimeInputMessage(message=data, message_index=index, eof=self._is_closed),
-                ),
-            )
+        try:
+            if self._object_type == "sandbox":
+                await retry_transient_errors(
+                    self._client.stub.SandboxStdinWrite,
+                    api_pb2.SandboxStdinWriteRequest(
+                        sandbox_id=self._object_id, index=index, eof=self._is_closed, input=data
+                    ),
+                )
+            else:
+                await retry_transient_errors(
+                    self._client.stub.ContainerExecPutInput,
+                    api_pb2.ContainerExecPutInputRequest(
+                        exec_id=self._object_id,
+                        input=api_pb2.RuntimeInputMessage(message=data, message_index=index, eof=self._is_closed),
+                    ),
+                )
+        except GRPCError as exc:
+            if exc.status == Status.FAILED_PRECONDITION:
+                raise ValueError("I/O operation on closed file.")
```

Comment:
Would still prefer if the client relied on the message from the server because in the future there could be multiple ways of receiving a `FAILED_PRECONDITION`.

Reply:
ok ser

Reply:
Wait can I convince you otherwise on this - I want to match the exact error that Python throws on close FDs for close API compatibility, but I don't think it makes sense for the RPC to throw this. I think the separation is valuable here, for the same reason that we don't throw FAILED_PRECONDITION back to the user.

Reply:
ok boss you convinced me